vcr/holder/oidc4vci_wallet.go

```diff
@@ -0,0 +1,178 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ */
+
+package holder
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/go-did/vc"
+	"github.com/nuts-foundation/nuts-node/audit"
+	"github.com/nuts-foundation/nuts-node/core"
+	"github.com/nuts-foundation/nuts-node/crypto"
+	"github.com/nuts-foundation/nuts-node/vcr/log"
+	"github.com/nuts-foundation/nuts-node/vcr/oidc4vci"
+	vcrTypes "github.com/nuts-foundation/nuts-node/vcr/types"
+	vdr "github.com/nuts-foundation/nuts-node/vdr/types"
+)
+
+// OIDCWallet is the interface for the credential wallet supporting OpenID4VCI.
+type OIDCWallet interface {
+	// Metadata returns the OAuth2 client metadata for the wallet.
+	Metadata() oidc4vci.OAuth2ClientMetadata
+	// HandleCredentialOffer handles a credential offer from an issuer.
+	// It will try to retrieve the offered credential and store it.
+	HandleCredentialOffer(ctx context.Context, offer oidc4vci.CredentialOffer) error
+}
+
+var nowFunc = time.Now
+var _ OIDCWallet = (*wallet)(nil)
+
+// NewOIDCWallet creates an OIDCWallet that tries to retrieve offered credentials, to store it in the given credential store.
+func NewOIDCWallet(did did.DID, identifier string, credentialStore vcrTypes.Writer, signer crypto.JWTSigner, resolver vdr.KeyResolver, clientTimeout time.Duration) OIDCWallet {
+	return &wallet{
+		did:                 did,
+		identifier:          identifier,
+		credentialStore:     credentialStore,
+		signer:              signer,
+		resolver:            resolver,
+		clientTimeout:       clientTimeout,
+		issuerClientCreator: oidc4vci.NewIssuerAPIClient,
+	}
+}
+
+type wallet struct {
+	did                 did.DID
+	identifier          string
+	credentialStore     vcrTypes.Writer
+	signer              crypto.JWTSigner
+	resolver            vdr.KeyResolver
+	clientTimeout       time.Duration
+	issuerClientCreator func(ctx context.Context, httpClient *http.Client, credentialIssuerIdentifier string) (oidc4vci.IssuerAPIClient, error)
+}
+
+func (h wallet) Metadata() oidc4vci.OAuth2ClientMetadata {
+	return oidc4vci.OAuth2ClientMetadata{
+		CredentialOfferEndpoint: core.JoinURLPaths(h.identifier, "/wallet/oidc4vci/credential_offer"),
+	}
+}
+
+// HandleCredentialOffer handles a credential offer from an issuer.
+// Error responses on the Credential Offer Endpoint are not defined in the OpenID4VCI spec,
+// so these are inferred of whatever makes sense.
+func (h wallet) HandleCredentialOffer(ctx context.Context, offer oidc4vci.CredentialOffer) error {
+	// TODO: This check is too simplistic, there can be multiple credential offers,
+	//       but the wallet should only request the one it's interested in.
+	//       See https://github.com/nuts-foundation/nuts-node/issues/2049
+	if len(offer.Credentials) == 0 {
```

Comment:
If there are multiple offers, the wallet will only request 1 in retrieveCredential. So either this check should check for 1 or the retrieve must retrieve multiple.

Reply:
Made it check exactly 1